damnit/gui/main_window.py

```diff
@@ -260,17 +268,56 @@ def autoconfigure(self, path: Path):
         self.show_default_status_message()
         self.context_dir_changed.emit(str(path))
         self.launch_update_computed_vars()
+        self.start_watching_context_file()
 
-    def launch_update_computed_vars(self):
+    def launch_update_computed_vars(self, ctx_size_mtime=None):
+        # Triggered when we open a proposal & when saving the context file
         log.debug("Launching subprocess to read variables from context file")
+        self.updating_vars = True
+        # Store the size & mtime before processing the file: better to capture
+        # this just before a change and process the same version twice than
+        # just after & potentially miss a change.
+        self.vars_ctx_size_mtime = ctx_size_mtime or self.get_context_size_mtime()
         proc = QtCore.QProcess(parent=self)
         # Show stdout & stderr with the parent process
         proc.setProcessChannelMode(QtCore.QProcess.ProcessChannelMode.ForwardedChannels)
+        proc.finished.connect(self.done_update_computed_vars)
         proc.finished.connect(proc.deleteLater)
         proc.setWorkingDirectory(str(self.context_dir))
         proc.start(sys.executable, ['-m', 'damnit.cli', 'read-context'])
         proc.closeWriteChannel()
```

Comment:
```suggestion
    def launch_update_computed_vars(self, ctx_size_mtime=None):
        if not self.updating_vars:
            # Triggered when we open a proposal & when saving the context file
            log.debug("Launching subprocess to read variables from context file")
            self.updating_vars = True
            # Store the size & mtime before processing the file: better to capture
            # this just before a change and process the same version twice than
            # just after & potentially miss a change.
            self.vars_ctx_size_mtime = ctx_size_mtime or self.get_context_size_mtime()
            proc = QtCore.QProcess(parent=self)
            # Show stdout & stderr with the parent process
            proc.setProcessChannelMode(QtCore.QProcess.ProcessChannelMode.ForwardedChannels)
            proc.finished.connect(self.done_update_computed_vars)
            proc.finished.connect(proc.deleteLater)
            proc.setWorkingDirectory(str(self.context_dir))
            proc.start(sys.executable, ['-m', 'damnit.cli', 'read-context'])
            proc.closeWriteChannel()
```
Given that the method is called from different places and not all are guarded via the `self.updating_vars` variable do you think it would make sense to move the check in the method itself? I'm especially referring at the call that happens here: `self.context_saved.connect(self.launch_update_computed_vars)`. If you decide to accept this change there is also to remove the check at line 299 (`if (self.vars_ctx_size_mtime != size_mtime) and not self.updating_vars:`).

Reply:
Overlapping runs of the subprocess shouldn't actually cause any problems, I was just trying to avoid unnecessary runs. Now that we've moved the size+mtime update before we launch the subprocess, I think we may as well get rid of this `updating_vars` state, so there's one less thing to think about.

Reply:
The potential issue that I see would happen if the two calls to `launch_update_computed_vars` happen right to each other and pick different contents from the file. Then if the external processes happen in the reverse order compared to the calls that could mask/undo an update (I think). I reckon, though, that this would be even more unlikely than the one I've reported before (many things have to happen in a very peculiar way for this to happen if it is even possible).

Reply:
I see what you mean, but I think it's better to take our chances with that than to make this more complex to try to protect against it. A mutex in the GUI doesn't protect against another instance of the GUI running somewhere, or someone updating the variables from the command line.

Any inconsistencies should get sorted out the next time you save the context file or reprocess a run.

Reply:
Yeah, I think your argument is sound.